rfcs/0156-no-direct-pushes.md

```diff
@@ -0,0 +1,173 @@
+---
+feature: no-direct-pushes
+start-date: 2023-07-21
+author: Silvan Mosberger
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Require pull requests for all Nixpkgs commits.
+
+# Motivation
+[motivation]: #motivation
+
+There are currently 197 people with commit access to Nixpkgs, all of whom can push directly to any branch without a pull request.
+Such pushes generally do not notify anybody and do not trigger CI.
+
+This makes such commits susceptible to:
+- Be anonymous
+- Include malicious code
+- Be broken
+- Have poor code quality
+
+While requiring pull requests isn't a panacea, it does help by:
+- Notifying [code owners](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners)
+- Running CI (though there's no requirement for it to succeed, see [future work][future])
+- Tying commits to a GitHub account
+- Being discoverable in the pull request list
+
+# Detailed design
+[design]: #detailed-design
+
+Turn on GitHub's "Require a pull request before merging" [branch protection rule](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule#creating-a-branch-protection-rule) for all branches whose commits get propagated into channels.
+This includes:
+- `master`: Used for unstable channels and branched into new release branches
+- `release-*`: Used for stable channels
+
+Staging branches are intentionally not included, because they will already require a pull request when they inevitably need to get merged into one of the above branches.
+The same applies to similar long-term branches like `haskell-packages`.
+
+A NixOS GitHub organization owner needs to implement this change and should therefore review this proposal.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Direct pushes listing
+
+Out of the 112217 commits to master in the last year[^1], _at most_ 58 (0.0517%) of them were direct pushes.
+
+[^1]: Unix epoch 1658361600 to 1689897600
+
+To determine whether a commit was pushed directly, the GitHub API was queried for pull requests associated with that commit (see [`associatedPullRequests`](https://docs.github.com/en/graphql/reference/objects#commit)).
+If this list includes a merged pull request to the Nixpkgs master branch, the commit is known to be merged with a pull request.
```

Comment:
I would add a remark here that GitHub does allow merging PRs via a direct push of a clean-merge commit. (I cannot figure out from the documentation if it allows pushing an unclean merge, which would be an issue if we took «malicious code» part seriously)

Reply:
Will test this